### PR TITLE
Catching up Problem of Points

### PR DESCRIPTION
**Pull request by Jiashun97**
Thanks for your detailed feedback on my code which really helps make it clearer and more organized. I did the following changes:

1. I adapted all the help functions to camelCase and changed the name of "compare functions" to make them more informative.
2. I replicated the changes you did in the game of chance to the game of skill.
3. I fixed the bug in the credible interval.

I was wondering if you can have a look at it and do tell me if there are additional changes that needed to be made. @FBartos

### CODE REVIEW

**Review by FBartos — CHANGES_REQUESTED** [review]
Thanks @Jiashun97 , the code looks very good to me and it will be much easier to maintain. I suggested just one minor change that I forgot to mention in the previous PR.

Could you also add [unit tests](https://github.com/jasp-stats/jasp-desktop/blob/development/Docs/development/r-unit-test-guide.md) for those two analyses?

**Comment by FBartos** [comment]
Also, I discovered some edge cases that we should take care of.

1) Inconsistent labeling of "probability" - that was most likely introduced by me when editing the qml. I'm not sure whether `Pr` or `p` is more appropriate, maybe someone else can weigh in on this.
image

2) Setting the number of simulated games to 1 crashes the analysis. I suggest setting a minimum allowed number to 100 in the qml.
image

3) Setting points needed to win the game to 0 or 1 does not trigger the error message that one of the players already won the game.
image

File: R/LSgameofchance.R [hunk 1]
```diff
@@ -76,8 +76,8 @@ LSgameofchance   <- function(jaspResults, dataset, options, state = NULL){
   ## fill in the table and the plot
   if (nPlayers == 2 & max(xPoints) < winPoints){
```
[the remark this answers is not in the document]
Comment (Jiashun97) [reply R/LSgameofchance.R]: Sure, I have added the unit test and changed the symbols.

[the remark this answers is not in the document]
Comment (Jiashun97) [reply R/LSgameofchance.R]: 1. I would just leave `p()` there for all and they are easy to change if necessary.
2. I add an `nSims` check in the analysis instead of qml, which has more informative content via the error message.
3. I add another check to make sure `winPoints>=1` and fix the boundary issue when `winPoints=1`.

[the remark this answers is not in the document]
Comment (FBartos) [reply R/LSgameofchance.R]: 1 & 3 is fine

I think that it is better to restrict the nSims in qml - for consistency as it's done with bootstrapping in other analyses + the user doesn't need to see an error message if it's impossible to set it to bad values

**Comment by FBartos** [comment]
thanks, looks good to me!

@vandenman , do you want to re-review it after me? Or do we just merge it in?

**Review by FBartos — APPROVED** [review]

File: R/LSproblemofpointscommon.R [hunk 2]
```diff
@@ -1,227 +1,234 @@
...
-  dDirichletnmultinom <- function(y, a){
+## This is a function calculating the probability mass function 
+# of Dirichlet negative multinomial distribution
+
+## Input:
+# y: a vector containing the number of trials each player needs to win
+# a: a vector containing all the parameters for the Dirichlet multinomial distribution, indicating the competence of each player
+
+pDirichletNegMultinom <- function(y, a){
+  dDirichletNegMultinom <- function(y, a){
```
Comment (vandenman) [inline R/LSproblemofpointscommon.R:9]: Two things here.

1. Please don't define functions inside other functions, unless they are lambdas (e.g., `sapply(list, function(x) {...})`).
2. What are typical values for `y` and `a`? Because the code as written here will very quickly overflow to `Inf`. For example,
```r
y <- a <- 20:25
prod(gamma(y + a))
# Inf
sum(lgamma(y + a))
# 752.6915
```
when working with probability distributions it is common to do all work on the log scale to avoid overflowing or underflowing. In that case that would become:
```r
dDirichletNegMultinom2 <- function(y, a, log = FALSE) {
  
  ySum <- sum(y)
  aSum <- sum(a)
  
  lp1 <- lgamma(ySum) - (lgamma(y[1]) + sum(lgamma(y+1)) - lgamma(y[1]+1))
  lp2 <- lgamma(aSum) - sum(lgamma(a))
  lp3 <- sum(lgamma(y + a)) - lgamma(ySum + aSum)

  logDensity <- lp1 + lp2 + lp3
  if (log)
    return(logDensity)
  else
    return(exp(logDensity))
    
}
```
The original `dDirichletNegMultinom` underflows when called like `dDirichletNegMultinom(1:20, 1:20)` but `dDirichletNegMultinom2` does not. Likewise, you could compute `pCalculated` with the logSumExp trick, but it kinda depends on typical values of `y` and `a` whether this is useful.

File: R/LSproblemofpointscommon.R [hunk 3]
```diff
@@ -1,227 +1,234 @@
...
-    y <- matrix(ncol = length(k), nrow = prod(k)/k[1])
-    for (i in 1:length(x1[,1])){
-      for (j in 1:k[length(k)]){
-        y[(i-1)*(k[length(k)])+j,] <- append(x1[i,],j-1)
+    x1 <- combinations(k[1:(n-1)]) # recursion
+    x2 <- k[n]
+    sets <- matrix(ncol = n, nrow = prod(k)/k[1])
+    for (i in 1:length(x1[ ,1])){
+      for (j in 1:k[n]){
+        sets[(i-1)*(k[n])+j, ] <- append(x1[i, ],j-1)
```
Comment (vandenman) [inline R/LSproblemofpointscommon.R:52]:
```suggestion
        sets[(i-1)*(k[n])+j, ] <- c(x1[i, ],j-1)
```

File: R/LSproblemofpointscommon.R [hunk 4]
```diff
@@ -1,227 +1,234 @@
...
 
-compare_function1 <- function(p,m1,n1,t,simulation){
+compareChanceTwoPlayers <- function(p,m1,n1,t,simulation){
 
   # first estimating the probability that player 1 wins by simulating the game for a given number of times.
   m <- t - m1 # m: the number of success that player 1 should have to win the game
   n <- t - n1 # n: the number of success that player 2 should have to win the game
 
-  record_game <- vector()
+  recordGame <- vector()
```
Comment (vandenman) [inline R/LSproblemofpointscommon.R:66]:
```suggestion
  recordGame <- numeric(simulation)
```
or `integer(simulation)` if you prefer to be pedantic about the types.

File: R/LSproblemofpointscommon.R [hunk 5]
```diff
@@ -1,227 +1,234 @@
...
-    # if record_trial[1,] = 0, record 1,player 1 wins this single trial;
-    # if record_trial[2,] = 0, record 0, player 2 wins
-    record_game[i] <- record_trial[2, ]^record_trial[1, ]
+    # if recordTrial[1, ] = 0, record 1,player 1 wins this single trial;
+    # if recordTrial[2, ] = 0, record 0, player 2 wins
+    recordGame[i] <- recordTrial[2, ]^recordTrial[1, ]
 
   }
-  p_simulated <- sum(record_game)/simulation    # estimated probability that player 1 wins
+  pSimulated <- sum(recordGame)/simulation    # estimated probability that player 1 wins
```
Comment (vandenman) [inline R/LSproblemofpointscommon.R:80]: So we have
```r
while (recordTrial[1, ] * recordTrial[2, ] != 0) {
  recordTrial <- recordTrial - rmultinom(1, 1, c(p, 1 - p))
}
```
which implies that after the loop either `recordTrial[1, ] == 0` or `recordTrial[2, ] == 0`. The other value can be any integer right?

Slightly more readable than `recordGame[i] <- recordTrial[2, ]^recordTrial[1, ]` would be
`recordGame[i] <- if (recordTrial[2, ] == 0) 0 else 1`.

Also, isn't `pSimulated <- sum(recordGame)/simulation` equal to `pSimulated <- mean(recordGame)`?

File: R/LSproblemofpointscommon.R [hunk 6]
```diff
@@ -1,227 +1,234 @@
...
-compare_function2 <- function(k1, t, p, simulation){
+compareChanceNPlayers <- function(k1, t, p, simulation){
   # first estimating the probability that player 1 wins
   k <- t-k1 # k: the vector regarding the number of more successes that player 1, 2, ..., n should have to win the game
 
   # normalize the probability if they do not sum to one
   p <- p/sum(p)
 
-  record_game <- vector()# initializing the vector to record the results of all the games
+  recordGame <- vector()# initializing the vector to record the results of all the games
```
Comment (vandenman) [inline R/LSproblemofpointscommon.R:103]: preallocating the data is typically a good idea
```suggestion
  recordGame <- numeric(simulation)# initializing the vector to record the results of all the games
```

**Review by vandenman — CHANGES_REQUESTED** [review]
some minor comments, see below.

File: R/LSproblemofpointscommon.R (hunk 2 above)
Comment (Jiashun97) [reply R/LSproblemofpointscommon.R:9]: Great point! Though typicall `y` and `a` are vectors containing small numbers with length 2 or 3, it is better to use the log scale to avoid overflowing to `Inf`.

File: R/LSproblemofpointscommon.R (hunk 5 above)
Comment (Jiashun97) [reply R/LSproblemofpointscommon.R:80]: Yes, this makes it much more readable.

File: R/LSproblemofpointscommon.R (hunk 6 above)
Comment (Jiashun97) [reply R/LSproblemofpointscommon.R:103]: I also made the same changes for the other two functions.

File: R/LSproblemofpointscommon.R (hunk 2 above)
Comment (Jiashun97) [reply R/LSproblemofpointscommon.R]: I just moved `dDirichletNegMultinom` out of `pDirichletNegMultinom`

**Comment by FBartos** [comment]
And just a minor comment for the next time. It's usually a good practice to name the commits to corresponds to the changes being made (or even add a short description). It makes it much easier to track commits that changed a specific functionality is something goes south down the line.

**Comment by Jiashun97** [comment]
> And just a minor comment for the next time. It's usually a good practice to name the commits to corresponds to the changes being made (or even add a short description). It makes it much easier to track commits that changed a specific functionality is something goes south down the line.

Yes, it makes sense. I will try to add sth in the commit description.

**Review by vandenman — APPROVED** [review]
